base/dataset_prepare.go

```diff
@@ -145,46 +145,66 @@ func InferValues(pro *profile.Profile, ds *dataset.Dataset) error {
 	// if we don't have a structure or schema then attempt to determine one
 	body := ds.BodyFile()
 	if body != nil && (ds.Structure == nil || ds.Structure.Schema == nil) {
-		// use a TeeReader that writes to a buffer to preserve data
-		buf := &bytes.Buffer{}
-		tr := io.TeeReader(body, buf)
-		var df dataset.DataFormat
-
-		df, err := detect.ExtensionDataFormat(body.FileName())
-		if err != nil {
-			log.Debug(err.Error())
-			err = fmt.Errorf("invalid data format: %s", err.Error())
+		if err := InferStructure(ds); err != nil {
 			return err
 		}
+	}
 
-		guessedStructure, _, err := detect.FromReader(df, tr)
-		if err != nil {
-			log.Debug(err.Error())
-			err = fmt.Errorf("determining dataset structure: %s", err.Error())
-			return err
-		}
+	if ds.Transform != nil && ds.Transform.ScriptFile() == nil && ds.Transform.IsEmpty() {
+		ds.Transform = nil
+	}
 
-		// attach the structure, schema, and formatConfig, as appropriate
-		if ds.Structure == nil {
-			ds.Structure = guessedStructure
-		}
-		if ds.Structure.Schema == nil {
-			ds.Structure.Schema = guessedStructure.Schema
-		}
-		if ds.Structure.FormatConfig == nil {
-			ds.Structure.FormatConfig = guessedStructure.FormatConfig
-		}
+	return nil
+}
 
-		// glue whatever we just read back onto the reader
-		// TODO (b5)- this may ruin readers that transparently depend on a read-closer
-		// we should consider a method on qfs.File that allows this non-destructive read pattern
-		ds.SetBodyFile(qfs.NewMemfileReader(body.FileName(), io.MultiReader(buf, body)))
+// InferStructure infers the Structure field of the dataset, guarenteeing
+// that the structure will contain a Format, FormatConfig, and Schema, based
+// on the given dataset body. It will not write over any Structure, Format,
+// FormatConfig, or Schema that already exists.
+func InferStructure(ds *dataset.Dataset) error {
```

Comment:
This is awesome! A super helpful function to have, I bet this can also be used to solve https://github.com/qri-io/qri/issues/1532 pretty easily.

Looking at https://github.com/qri-io/qri/blob/master/fsi/body.go, it seems fsi does some extra tricks, by using `Seek`, which is can do as it knows it's working with local files. If we use `InterStructure` to also handle "get body" for fsi, we can either drop that extra efficiency gain, or add some dynamic interface checking to this function.